### PR TITLE
Update activities controller to use create_async

### PR DESCRIPTION
**Pull request by philbogle**
Change the activities controller you to use create_async!, except when saving to the gallery, when we continue to use create! due to the need to access the activity_id.

### CODE REVIEW

**Comment by philbogle** [comment]
@laurelfan Here's the conditional async create for activities.

**Comment by laurelfan** [comment]
Does this code path still work:

https://github.com/code-dot-org/code-dot-org/blob/a91b1a18c87c71230fe9033f0fdc6fc7fa264e05/dashboard/app/controllers/application_controller.rb#L174

(There are two ways GalleryActivities are created -- some automatically for "autosaved" levels and some explicitly by clicking the "Save to Gallery" button, which is generated by blockly using the save_to_gallery_url that dashboard passes to it in the response)

**Comment by philbogle** [comment]
Ready for another look.

@cpirich helped me with an elegant solution for the freeplay level problem.  There is a special testResult code (ActivityConstants::FREE_PLAY_RESULT = 30) which is posted by the javascript for the freeplay levels.  So even if Ruby doesn't know that a particular level is freeplay the testResult code tells us when we need to save synchronously.

**Comment by philbogle** [comment]
Also added a number of unit tests with async writes enabled and disabled, and one which verifies the async write is really queued and the applied correctly from the the queue when handled.

File: dashboard/app/controllers/activities_controller.rb [hunk 1]
```diff
@@ -149,7 +163,7 @@ def track_progress_for_user
     end
 
     # blockly sends us 'undefined', 'false', or 'true' so we have to check as a string value
-    if params[:save_to_gallery] == 'true' && @level_source_image && solved
+    if  params[:save_to_gallery] == 'true' && @level_source_image && solved
```
Comment (laurelfan) [inline dashboard/app/controllers/activities_controller.rb]: extra space

**Comment by laurelfan** [comment]
LGTM! I'm glad you guys figured out the FREE_PLAY thing so we don't have to do a bunch of duplicate config nonsense.

**Pull request merged**